### DISCUSSION TITLE
SD card Initialization stuck in check_device_status()

### DISCUSSION THREAD

**Discussion by alee1234567** [polarfire-soc/polarfire-soc-documentation#146] [Bare metal embedded software] [answered]
Hello I started another discussion about the DDRAM a week ago, but I've sine decided to put that on pause. 

I'm testing the SD card an am using the mss_mmc driver files and the sd card from the polarfire baremetal library (mpfs-emmc-sd-write-read).

It is able to send commands and receive responses, but it is stuck at line 3499 of mss_mmc.c. This line goes to check_device_status() and performs do while loop as seen in this screenshot:
![image](https://user-images.githubusercontent.com/109090085/180317583-82b8bf69-3a7a-4fce-96dc-6085c2fb7f74.png)
It never leaves the do while loop as srs9 & SRS9_DAT0_SIGNAL_LEVEL never equal MSS_CLEAR. 

  
![image](https://user-images.githubusercontent.com/109090085/180317439-0a179b8d-7efe-4564-97fb-d857e1e28006.png)

**Comment by hughbreslin** [comment, marked as answer]
Hi @alee1234567 

I think based on the previous discussion this is a custom board correct? Can I check what drive strength you have set for your SD MSSIOs? In my case they're connected to BANK4:

![image](https://user-images.githubusercontent.com/44862881/180422532-7683eef9-c416-4e1a-900b-02ce55cfefa6.png)

Our default value for the Icicle Kit is 8mA but we'll be updating this to 12mA in a future release, this may resolve your issue if you can try the change?

We also have some updates planned for the driver too, if you could let me know on the above lets see! 🙂

**Reply by alee1234567** [reply]
It was at 8 mA, but I changed it to 12 mA, and it did fix the issue.

**Reply by hughbreslin** [reply]
Excellent :) The reason for this is that for the default speeds 8mA is acceptable, for higher speeds 12mA is recommended. This is noted in section 6.1.5 of the [PolarFire SoC Advanced Datasheet](https://ww1.microchip.com/downloads/aemDocuments/documents/FPGA/ProductDocuments/DataSheets/microsemi-microchip_polarfire_soc_advanced_datasheet_10_12-5-19.pdf) and the MSS configurator is also going to be updated with profiles for the different configurations to make it easier for people. 

      The following tables describe SD_SDIO. The test conditions for SD/SDIO standard mode (default speed
      mode) use an 8 mA drive strength, fast slew rate, and a 30 pF load. For SD/SDIO high-speed mode, the test
      conditions use a 12 mA drive strength, fast slew rate, and a 30 pF load. For other SD/SDIO high-speed
      modes, the test conditions use a 12 mA drive strength, fast slew rate, and a 15 pF load.
